src/mcmc/gibbs.jl

```diff
@@ -1,251 +1,528 @@
+_maybevec(x::VarName) = [x]
 
-# Initialize the Gibbs sampler.
 function DynamicPPL.initialstep(
-    rng::AbstractRNG, model::Model, spl::Sampler{<:Gibbs}, vi::AbstractVarInfo; kwargs...
+    rng::Random.AbstractRNG,
+    model::DynamicPPL.Model,
+    spl::DynamicPPL.Sampler{<:Gibbs},
+    vi_base::DynamicPPL.AbstractVarInfo;
+    initial_params=nothing,
+    kwargs...,
 )
-    # TODO: Technically this only works for `VarInfo` or `ThreadSafeVarInfo{<:VarInfo}`.
-    # Should we enforce this?
-
-    # Create tuple of samplers
-    algs = spl.alg.algs
-    i = 0
-    samplers = map(algs) do alg
-        i += 1
-        if i == 1
-            prev_alg = algs[end]
-        else
-            prev_alg = algs[i - 1]
-        end
-        rerun = gibbs_rerun(prev_alg, alg)
-        selector = DynamicPPL.Selector(Symbol(typeof(alg)), rerun)
-        Sampler(alg, model, selector)
+    alg = spl.alg
+    varnames = alg.varnames
+    samplers = alg.samplers
+
+    # Run the model once to get the varnames present + initial values to condition on.
+    vi = DynamicPPL.VarInfo(rng, model)
+    if initial_params !== nothing
+        vi = DynamicPPL.unflatten(vi, initial_params)
     end
 
-    # Add Gibbs to gids for all variables.
-    for sym in keys(vi.metadata)
-        vns = getfield(vi.metadata, sym).vns
-
-        for vn in vns
-            # update the gid for the Gibbs sampler
-            DynamicPPL.updategid!(vi, vn, spl)
+    # Initialise each component sampler in turn, collect all their states.
+    states = []
+    for (varnames_local, sampler_local) in zip(varnames, samplers)
+        varnames_local = _maybevec(varnames_local)
+        # Get the initial values for this component sampler.
+        initial_params_local = if initial_params === nothing
+            nothing
+        else
+            DynamicPPL.subset(vi, varnames_local)[:]
+        end
 
-            # try to store each subsampler's gid in the VarInfo
-            for local_spl in samplers
-                DynamicPPL.updategid!(vi, vn, local_spl)
-            end
+        # Construct the conditioned model.
+        model_local, context_local = make_conditional(model, varnames_local, vi)
+
+        # Take initial step.
+        _, new_state_local = AbstractMCMC.step(
+            rng,
+            model_local,
+            sampler_local;
+            # FIXME: This will cause issues if the sampler expects initial params in unconstrained space.
+            # This is not the case for any samplers in Turing.jl, but will be for external samplers, etc.
+            initial_params=initial_params_local,
+            kwargs...,
+        )
+        new_vi_local = varinfo(new_state_local)
+        # TODO(mhauru) Remove the below loop once samplers no longer depend on selectors.
+        # For some reason not having this in place was causing trouble for ESS, but not for
+        # other samplers. I didn't get to the bottom of it.
+        for vn in keys(new_vi_local)
```

Comment:
Is this just because you haven't updated the `ESS` sampler yet? Looking at it's constructor, it has one that takes in a `Symbol`, thus it will have a `space` that is only that symbol(s). With the new `Gibbs`, none of the `InferenceAlgorithm`s should have a symbol attached, right?

Reply:
Ah, I see that we aren't actually altering this 😕 

My suggestion here is to do the following:
1. Make a function called `drop_space(::InferenceAlgorithm)` which we implement for our existing samplers. This should a new instance of the alg but _without_ any `space`.
2. Call `drop_space` on the algs provided to `Gibbs`.

If we do this, then we can keep the old interface still functioning but without relying on `space` at all internally:)

Reply:
For example, for `ESS`, we just do
```julia
drop_space(::ESS) = ESS()
```
which would then drop the `space` in `ESS{space}`.

Reply:
I like the idea of `drop_space`. Will have to see that that doesn't cause trouble with e.g. MH, which I think uses spaces to specify proposal distributions, or something like that. Not having to edit those sorts of internals of samplers in the same PR was the reason why I tried to just ignore, rather than suppress, the old space `Selector` mechanism.

Reply:
Done.